services/src/main/java/org/keycloak/utils/GroupUtils.java

```diff
@@ -1,56 +1,80 @@
 package org.keycloak.utils;
-
-import java.util.Collections;
-import java.util.stream.Collectors;
-
-import org.keycloak.common.Profile;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
 import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
 import org.keycloak.models.utils.ModelToRepresentation;
 import org.keycloak.representations.idm.GroupRepresentation;
 import org.keycloak.services.resources.admin.permissions.GroupPermissionEvaluator;
 
-public class GroupUtils {
-    // Moved out from org.keycloak.admin.ui.rest.GroupsResource
-    public static GroupRepresentation toGroupHierarchy(GroupPermissionEvaluator groupsEvaluator, GroupModel group, final String search, boolean exact, boolean lazy) {
-        return toGroupHierarchy(groupsEvaluator, group, search, exact, true, lazy);
-    }
 
-    public static GroupRepresentation toGroupHierarchy(GroupPermissionEvaluator groupsEvaluator, GroupModel group, final String search, boolean exact, boolean full, boolean lazy) {
-        GroupRepresentation rep = ModelToRepresentation.toRepresentation(group, full);
-        if (!lazy) {
-            rep.setSubGroups(group.getSubGroupsStream().filter(g ->
-                    groupMatchesSearchOrIsPathElement(
-                            g, search
-                    )
-            ).map(subGroup ->
-                    ModelToRepresentation.toGroupHierarchy(
-                            subGroup, full, search, exact
-                    )
 
-            ).collect(Collectors.toList()));
-        } else {
-            rep.setSubGroups(Collections.emptyList());
-        }
+public class GroupUtils {
 
-        if (Profile.isFeatureEnabled(Profile.Feature.ADMIN_FINE_GRAINED_AUTHZ)) {
-            setAccess(groupsEvaluator, group, rep);
-        }
+    /**
+     * This method takes the provided groups and attempts to load their parents all the way to the root group while maintaining the hierarchy data
+     * for each GroupRepresentation object. Each resultant GroupRepresentation object in the stream should contain relevant subgroups to the originally
+     * provided groups
+     * @param session The active keycloak session
+     * @param realm The realm to operate on
+     * @param groups The groups that we want to populate the hierarchy for
+     * @return A stream of groups that contain all relevant groups from the root down with no extra siblings
+     */
+    public static Stream<GroupRepresentation> populateGroupHierarchyFromSubGroups(KeycloakSession session, RealmModel realm, Stream<GroupModel> groups, boolean full, GroupPermissionEvaluator groupEvaluator) {
+        Map<String, GroupRepresentation> groupIdToGroups = new HashMap<>();
+        groups.forEach(group -> {
```

Comment:
I have seen costly methods like this being run multiple times during a same session (KeycloakSession). Would mind checking if it happens with this method so that we can either change the code to execute only once or cache the result into the current session?

Reply:
Assuming my understanding of a KeycloakSession is accurate (I believe that the session is active for all users across the entire runtime of the server and that it doesn't represent a single HTTP transaction's contextual data?) then this method definitely gets run multiple times per session as it takes search results and constructs the tree based off of it. 

This method is, however, faster than the previous hierarchy algorithm. The previous algorithm would take the stream of top level groups, iterate through all of them until it found a matching group, load that group's entire subgroup list, and then repeat the algorithm on the next layer of the tree. 

For a setups with groups A,B,C where
A and B are top level groups 
C is a subgroup of B 
{ A, B { C } } 

A search that should return group C and its parent (group B) would end up also processing and then discarding Group A as well. This scaled out both vertically and horizontally for each matching group and each layer of the tree. 

The new algorithm at least ensures that only groups C and B and processed. 

As far as caching the results go, in other parts of the codebase I've noticed that when search is involved the cache tends to be bypassed in favor of getting the results directly from the database. Would this follow that same design mindset?

Reply:
@Redhat-Alice The `KeycloakSession` is created on a per-request basis. My point was to use an attribute in the current `KeycloakSession` to store the results when running that code for the first time so that we can skip running it during the session/request lifetime.

If this code is only run once during a request lifetime, please ignore me :) Otherwise, I think it is worth including this "caching".

I agree your code should perform better.

Reply:
We should be good here then! This code is only run right before returning the response!